File: leetcode/7_sorting/179_Largest_Number.py
"""
Given a list of non-negative integers nums, arrange them such that they form the largest number and return it.

Since the result may be very large, so you need to return a string instead of an integer.



Example 1:

Input: nums = [10,2]
Output: "210"

Example 2:

Input: nums = [3,30,34,5,9]
Output: "9534330"



Constraints:

    1 <= nums.length <= 100
    0 <= nums[i] <= 109
"""


class Solution:
    def largestNumber(self, nums: List[int]) -> str:
        _compare = lambda n1, n2: -1 if n1 + n2 > n2 + n1 else 1

        nums = list(map(str, nums))
        nums = sorted(nums, key=cmp_to_key(_compare))
        if nums[0] == "0":
            return "0"
        return "".join(nums)

File: leetcode/7_sorting/179_Largest_Number.go
import (
	"sort"
	"strconv"
	"strings"
)

func largestNumber(nums []int) string {
	numStrs := make([]string, len(nums))

	// Convert integers to strings
	for i, num := range nums {
		numStrs[i] = strconv.Itoa(num)
	}

	// Sort based on custom rules
	sort.Slice(numStrs, func(i, j int) bool {
		return numStrs[i]+numStrs[j] > numStrs[j]+numStrs[i]
	})

	// If the largest number is "0", return "0"
	if numStrs[0] == "0" {
		return "0"
	}

	// Join the sorted strings to form the largest number
	return strings.Join(numStrs, "")
}
